include/lbann/callbacks/checkpoint.hpp

```diff
@@ -177,9 +177,18 @@ class checkpoint : public callback_base {
   }
 
   bool need_checkpoint(model *m, callback_phase phase);
-  std::string find_latest_checkpoint(model *m, std::string& latest_file, execution_mode& mode, size_t &epoch, size_t& step, int& shared);
-  bool open_latest_checkpoint(model *m,
+  std::string find_latest_checkpoint(lbann_comm& comm,
+                                     const std::string& trainer_name,
+                                     const std::string& alg_name,
+                                     std::string& latest_file,
+                                     execution_mode& mode,
+                                     size_t &epoch,
+                                     size_t& step,
+                                     int& shared);
```

Comment:
I'm concerned about this function from a semantics/documentation point of view. Some questions that come up include: 
1. Why does this function modify `comm`? It should be `const lbann_comm&` (yes, I know that there are some data members that aren't -- but should be -- marked `mutable` so the class is not `const`-correct).
2. What does the `std::string` returned by this mean? Is it the whole checkpoint in binary? Is it a directory name? Is it a file name? Is it a random series of bytes?
3. Is `latest_file` supposed to be an in/out parameter or just an output? Same for `mode`, `epoch`, `step`, and `shared`?
4. Is `shared` supposed to be `bool`?
I see in the implementation that the output string is a directory name. `latest_file` is only used by the trainer master rank. `mode`, `epoch`, `step`, and `shared` are in/out arguments on rank 0 of the trainer comm but output arguments on other ranks. And `shared` is indeed a `bool` in disguise. This should all be documented.

Moreover, this is a lot of POD floating around -- perhaps it's better to wrap this up into a struct with all the important info:
```
struct ckpt_info {
  std::string latest_file;
  std::string dir_name;
  execution_mode mode;
  size_t epoch,
  size_t step,
  bool shared
};
/** @brief Fill in metadata for the latest checkpoint.
 *   // Other doc here //
 *  @param[in,out] info The `ckpt_info` object describing the latest known checkpoint. Input on trainer master, output on other.
 */
void find_latest_checkpoint(const lbann_comm&, const std::string&, const std::string&, ckpt_info& info) const;
```
Also, if this is going to be a member function, it should be `const`.

Reply:
I fixed up several of these issues.